Integration of Victron SmartShunt via VE.Direct

philippsandhaus opened the pull request:
This a first draft of my integration as discussed in https://github.com/helgeerbe/OpenDTU-OnBattery/issues/205. As I did quite a lot of restructuring and I am not that confident in C++ I am marking this as draft and hereby ask for feedback from @schlimmchen and/or @helgeerbe. I rebased the current status on top of the development branch.

<img width="1386" alt="Bildschirmfoto 2023-09-17 um 22 38 20" src="https://github.com/helgeerbe/OpenDTU-OnBattery/assets/3939346/217105c1-c5c8-4a1f-8494-65aafd43e19c">
<img width="1368" alt="Bildschirmfoto 2023-09-17 um 22 08 23" src="https://github.com/helgeerbe/OpenDTU-OnBattery/assets/3939346/8f7a92c3-0e82-46bf-b8a7-91adbafd7529">

helgeerbe commented:
@philippsandhaus good job! I just looked at the frame handler. I would propose to use a `VeDirectController`base class and derive 'veDirectShuntController`and `VeDirectMpptController`from it. 

Advantage is, that you can do the generic stuff like init of the HardwareSerial in the base class and you have to implement it only once. Do only device specific in the DeviceController class.

schlimmchen commented on webapp/src/locales/en.json:
```diff
@@ -608,6 +608,7 @@
         "Provider": "Data Provider",
         "ProviderPylontechCan": "Pylontech using CAN bus",
         "ProviderJkBmsSerial": "Jikong (JK) BMS using serial connection",
+        "ProviderVictron": "Victron SmartShunt using VE.Direct Interface",
```
This is missing in fr.json.

philippsandhaus replied:
Hmm, the whole batteryadmin section is missing for. I copied the complete section from en for now. I guess the whole file has to be translated at some point.

philippsandhaus commented on lib/VeDirectFrameHandler/VeDirectFrameHandler.h:
```diff
@@ -66,6 +52,9 @@ class MovingAverage {
         return static_cast<double>(_sum) / _count;
     }
 
+   
```
Noting ;-) I removed them.

philippsandhaus commented on include/BatteryStats.h:
```diff
@@ -98,3 +99,24 @@ class JkBmsBatteryStats : public BatteryStats {
+
+        void updateFrom(VeDirectShuntController::veShuntStruct const& shuntData);
+
+        void setSoC(uint8_t SoC) { _SoC = SoC; _lastUpdateSoC = millis(); }
```
Yes, removed it

philippsandhaus commented on include/BatteryStats.h:
```diff
@@ -98,3 +99,24 @@ class JkBmsBatteryStats : public BatteryStats {
+        void updateFrom(VeDirectShuntController::veShuntStruct const& shuntData);
+
+        void setSoC(uint8_t SoC) { _SoC = SoC; _lastUpdateSoC = millis(); }
+        void setLastUpdate(uint32_t ts) { _lastUpdate = ts; }
```
Yes, removed it.

philippsandhaus commented on include/BatteryStats.h:
```diff
@@ -98,3 +99,24 @@ class JkBmsBatteryStats : public BatteryStats {
+        void setSoC(uint8_t SoC) { _SoC = SoC; _lastUpdateSoC = millis(); }
+        void setLastUpdate(uint32_t ts) { _lastUpdate = ts; }
+
+        float _voltage; 
```
Yes, I did. Made them private now.

philippsandhaus commented on lib/VeDirectFrameHandler/VeDirectMpptController.cpp:
```diff
@@ -0,0 +1,161 @@
+	_vedirectSerial->begin(19200, SERIAL_8N1, rx, tx);
+    _vedirectSerial->flush();
+	VeDirectFrameHandler::init(msgOut, verboseLogging);
+	if (_verboseLogging) _msgOut->println("Finished init MPPTCOntroller");
```
Done

philippsandhaus commented on webapp/src/locales/de.json:
```diff
@@ -599,6 +599,7 @@
         "Provider": "Datenanbieter",
         "ProviderPylontechCan": "Pylontech per CAN-Bus",
         "ProviderJkBmsSerial": "Jikong (JK) BMS per serieller Verbindung",
+        "ProviderVictron": "Victron SmartShunt per VE.Direct Interface",
```
Done

philippsandhaus commented on src/VictronSmartShunt.cpp:
```diff
@@ -0,0 +1,39 @@
+            pin.battery_rx, pin.battery_tx);
+
+    if (pin.battery_rx < 0 || pin.battery_tx < 0) {
+        MessageOutput.println(F("[Pylontech] Invalid pin config"));
```
Adjusted ;-)

philippsandhaus commented on webapp/src/locales/en.json:
```diff
@@ -835,6 +836,10 @@
         "overVoltage": "Overvoltage",
-        "bmsInternal": "BMS internal"
+        "bmsInternal": "BMS internal",
+        "nrChargedCycles": "Number of charged cycles",
```
I just looked into the Victron App. It just calls it charge cycles (Ladezyklen). So I guess I just copy that.

philippsandhaus commented on webapp/src/locales/fr.json:
```diff
@@ -772,6 +772,10 @@
+        "nrChargedCycles": "Number of charged cycles",
+        "chargedEnergy": "Charged energy",
+        "dischargedEnergy": "Discharged energy",
+        "model": "Model"
```
It was in an intermediate version. I removed it now.

philippsandhaus commented on lib/VeDirectFrameHandler/VeDirectMpptController.h:
```diff
@@ -0,0 +1,45 @@
+
+private:
+
+    virtual void textRxEvent(char * name, char * value);
```
Thanks, this is exactly the kind of feedback I need. Forces me to understand the meanings of final, virtual, override etc. I guess final is the right one here.

philippsandhaus commented on lib/VeDirectFrameHandler/VeDirectShuntController.h:
```diff
@@ -0,0 +1,54 @@
+private:
+
+    virtual void textRxEvent(char * name, char * value);
+    void frameEndEvent(bool) override;                 // copy temp struct to public struct
```
Changed to final

philippsandhaus commented on lib/VeDirectFrameHandler/VeDirectShuntController.h:
```diff
@@ -0,0 +1,54 @@
+    
+private:
+
+    virtual void textRxEvent(char * name, char * value);
```
Changed to final

philippsandhaus commented:
Thanks for the review. I adjusted my code accordingly and also moved initialization of the HardwareSerial interface to the base class. As far as I know there are only 2 HardwareSerial interfaces on the ESP32, right? So if there are more than 2 serial devices, one would have to think about some kind of management and also use SoftwareSerial, I guess.

Yes, for me the MPPT interface still works as before.

schlimmchen commented on src/VictronSmartShunt.cpp:
```diff
@@ -0,0 +1,39 @@
+    auto tx = static_cast<gpio_num_t>(pin.battery_tx);
+    auto rx = static_cast<gpio_num_t>(pin.battery_rx);
+
+    VeDirectShunt.init(rx, tx, &MessageOutput, verboseLogging);
```
`_veDirectShunt = std::make_unique<VeDirectShuntController>()` goes here. consider removing the VeDirectShuntController::init() method and doing its work in the constructor instead.

helgeerbe commented:
I wouldn't mind, if we switch to Software Serial completely for ve.direct devices. Guess this makes the orchestration easier. @schlimmchen pointed out baudrate is moderate and the actual loop time is fast enough. 

Just thinking loud, if we are going to support also a second charger.

philippsandhaus commented:
`SoftwareSerial` and `HardwareSerial` both derive from `Stream`. I envision a class like `SerialProvider` which provides instances of `Stream` with for example `SerialProvider.getSerial(19200, SERIAL_8N1, rx, tx)`. 'SerialProvider' could then make the decision to hand an instance of `HardwareSerial` or `SoftwareSerial` based on baudrate and/or availability of hardware serial ports. But I guess the tricky parts are the details like how to get notified when a Hardware Serial port is freed up.

schlimmchen commented:
@philippsandhaus Interesting thoughts! How about you open up a new issue where we can discuss this (rather than being buried in this PR)?

schlimmchen commented:
@philippsandhaus I see there are values in `veShuntStruct` that are of type double. However, all values are always converted using `atoi`. I think it's fine to save the values as they were on the serial line, i.e., integers, and to do the scaling later -- even though this is different than in the MPPT counterpart. You do the scaling in `VictronSmartShuntStats::updateFrom`. If so, shouldn't the values in the `veShuntStruct` be all `int32_t`?

schlimmchen commented:
@philippsandhaus Feel free to cherry pick from https://github.com/schlimmchen/OpenDTU-OnBattery/commits/devel for this PR -- if you like. And except for the SoftwareSerial commit, that's a work in progress and requires testing.

philippsandhaus commented on lib/VeDirectFrameHandler/VeDirectFrameHandler.cpp:
```diff
@@ -81,10 +78,11 @@ void VeDirectFrameHandler::setVerboseLogging(bool verboseLogging)
 }
 
-void VeDirectFrameHandler::init(int8_t rx, int8_t tx, Print* msgOut, bool verboseLogging)
+void VeDirectFrameHandler::init(Print* msgOut, bool verboseLogging, uint16_t rx, uint16_t tx, uint16_t hwSerialPort)
```
Cherry picked your changes

philippsandhaus commented on lib/VeDirectFrameHandler/VeDirectFrameHandler.cpp:
```diff
@@ -81,10 +78,11 @@ void VeDirectFrameHandler::setVerboseLogging(bool verboseLogging)
 {
-	VedirectSerial.begin(19200, SERIAL_8N1, rx, tx);
-	VedirectSerial.flush();
+	_vedirectSerial = new HardwareSerial(hwSerialPort);
```
Cherry picked your changes

philippsandhaus commented on lib/VeDirectFrameHandler/VeDirectFrameHandler.cpp:
```diff
@@ -81,10 +78,11 @@ void VeDirectFrameHandler::setVerboseLogging(bool verboseLogging)
-	VedirectSerial.flush();
+	_vedirectSerial = new HardwareSerial(hwSerialPort);
+	_vedirectSerial->begin(19200, SERIAL_8N1, rx, tx);
+    _vedirectSerial->flush();
```
Cherry picked your changes

philippsandhaus commented:
Great, I have cherry picked most of your adjustments. I also added handling of alarm cases for the live view. At least for me this PR would now be complete.

schlimmchen commented:
> At least for me this PR would now be complete.

Affirmative.

helgeerbe commented:
Cool. I could merge it into development. But I'm away from keyboard next week. So I would keep it in development for one additional week for further testing, before I release it. Are you fine with that?

swingstate commented:
I'll install a SmartShunt 500A tomorrow morning and was assuming you're going to release, however no worries - I'll compile development and flash that to my DTU for testing. Will report any findings and if no issue report back a "no issues"

philippsandhaus commented:
> Cool. I could merge it into development. But I'm away from keyboard next week. So I would keep it in development for one additional week for further testing, before I release it. Are you fine with that?

Sure, no problem.

swingstate commented:
guess now it is. I'll compile and flash my ESP tomorrow morning.

helgeerbe commented:
It's in development now. And it's on my OpenDTU.
Happy testing!

helgeerbe commented:
I removed the line. From my point of view. You should do the build locally. 
Action is only used to ensure, that every commit to development results in code that is compilable.

swingstate commented:
@helgeerbe I have the code on my ESP now and so far things look good, SOC is being displayed. I have one unclarity which I shared in the longer issue thread #205.
